webapp/src/main/java/com/box/l10n/mojito/service/translationkit/TranslationKitService.java

```diff
@@ -322,4 +304,95 @@ public void updateStatistics(Long translationKitId, Set<String> notFoundTextUnit
         return map;
     }
 
+    /**
+     * Gets the list of {@link TextUnitDTO}s to generate a
+     * {@link TranslationKit}
+     *
+     * @param translationKitId {@link TranslationKit}'s id
+     * @param type
+     *
+     * @return the list of {@link TextUnitDTO}s to generate a
+     * {@link TranslationKit}
+     */
+    public List<TextUnitDTO> getTextUnitDTOsForTranslationKitWithInheritance(Long translationKitId) {
+
+        TranslationKit translationKit = translationKitRepository.findOne(translationKitId);
+        Long repositoryId = translationKit.getDrop().getRepository().getId();
+        Long localeId = translationKit.getLocale().getId();
+        Stack<Long> localeIds = getLocaleInheritanceStack(repositoryId, localeId);
+
+        Map<Long, TextUnitDTO> mergedTextUnitDTOs = new TreeMap<>();
+        List<TextUnitDTO> textUnitDTOs = null;
+        while (!localeIds.empty()) {
+            Long currentLocaleId = localeIds.pop();
+            if (currentLocaleId == localeId) {
+                // child locale
+                textUnitDTOs = getTextUnitDTOsForTranslationKit(repositoryId, currentLocaleId, StatusFilter.REVIEW_NEEDED);
+                logger.info("child locale {} has {} text units for review", currentLocaleId, textUnitDTOs.size());
+            } else {
+                // parent locale
+                textUnitDTOs = getTextUnitDTOsForTranslationKit(repositoryId, currentLocaleId, StatusFilter.TRANSLATED_AND_NOT_REJECTED);
+                logger.info("parent locale {} has {} text units that are translated and not rejected", currentLocaleId, textUnitDTOs.size());
+            }
+            for (TextUnitDTO textUnitDTO : textUnitDTOs) {
+                mergedTextUnitDTOs.put(textUnitDTO.getTmTextUnitId(), textUnitDTO);
+            }
+        }
+        return new ArrayList<TextUnitDTO>(mergedTextUnitDTOs.values());
+    }
+
+    /**
+     * Returns {@link Stack} of locale Ids with ancestor on the top.
+     *
+     * @param repositoryId
+     * @param localeId
+     * @return
+     */
+    private Stack<Long> getLocaleInheritanceStack(Long repositoryId, Long localeId) {
+        Stack<Long> localeInheritance = new Stack<>();
+        RepositoryLocale repositoryLocale = repositoryLocaleRepository.findByRepositoryIdAndLocaleId(repositoryId, localeId);
+        while (repositoryLocale != null) {
+            localeInheritance.push(repositoryLocale.getLocale().getId());
+            repositoryLocale = repositoryLocale.getParentLocale();
+        }
+        return localeInheritance;
+    }
+
+    private List<TextUnitDTO> getTextUnitDTOsForTranslationKit(Long repositoryId, Long localeId, StatusFilter statusFilter) {
+
+        int retry = 1;
+        int maxRetry = 3;
+        List<TextUnitDTO> textUnitDTOs = null;
+
+        while (retry <= maxRetry) {
+            try {
```

Comment:
wondering why we had this retry logic, seems useless. try to remove it?

Reply:
Yeah, let's give a try.